File: Object-C/BlackFireIM/Example/ChatDetail/ChatRoom/BFChatRoomViewController.m
//
//  BFChatRoomViewController.m
//  BlackFireIM
//
//  Created by benny wang on 2021/10/29.
//

#import "BFChatRoomViewController.h"
#import "MSIMSDK-UIKit.h"
#import "YBImageBrowser.h"
#import "YBIBVideoData.h"
#import "BFChatRoomEditController.h"
#import "BFEditTodInfoController.h"


@interface BFChatRoomViewController ()<MSChatRoomControllerDelegate>

@property(nonatomic,strong) MSChatRoomController *chatController;

@property(nonatomic,strong) MSGroupInfo *roomInfo;

@end

@implementation BFChatRoomViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.view.backgroundColor = [UIColor d_colorWithColorLight:RGB(240, 240, 240) dark:TController_Background_Color_Dark];
    [self.navView.rightButton setImage:[UIImage d_imageWithImageLight:@"more_icon" dark:@"more_icon_white"] forState:UIControlStateNormal];
    
    self.chatController = [[MSChatRoomController alloc]init];
    self.chatController.delegate = self;
    self.chatController.roomInfo = self.roomInfo;
    [self addChildViewController:self.chatController];
    [self.view addSubview:self.chatController.view];
    
    self.navView.navTitleL.text = self.roomInfo.room_name;
    [self addNotifications];
}

- (void)addNotifications
{
    WS(weakSelf)
    [[NSNotificationCenter defaultCenter] addObserverForName:MSUIKitNotification_EnterChatroom_success object:nil queue:NSOperationQueue.mainQueue usingBlock:^(NSNotification * _Nonnull note) {
        [weakSelf enterChatRoom: note];
    }];
    [[NSNotificationCenter defaultCenter] addObserverForName:MSUIKitNotification_ChatRoomConv_update object:nil queue:NSOperationQueue.mainQueue usingBlock:^(NSNotification * _Nonnull note) {
        [weakSelf chatRoomEvent: note];
    }];
    [[NSNotificationCenter defaultCenter] addObserverForName:MSUIKitNotification_ChatRoom_Event object:nil queue:NSOperationQueue.mainQueue usingBlock:^(NSNotification * _Nonnull note) {
        [weakSelf chatRoomEvent: note];
    }];
    [[NSNotificationCenter defaultCenter] addObserverForName:@"kChatRoomTipsDidTap" object:nil queue:NSOperationQueue.mainQueue usingBlock:^(NSNotification * _Nonnull note) {
        [weakSelf chatRoomTipsDidTap];
    }];
}

- (void)dealloc
{
    [[NSNotificationCenter defaultCenter]removeObserver:self];
}

- (MSGroupInfo *)roomInfo
{
    return MSChatRoomManager.sharedInstance.chatroomInfo;
}

/// 聊天室设置界面
- (void)nav_rightButtonClick
{
    if (self.roomInfo) {
        [self.view endEditing:YES];
        BFChatRoomEditController *vc = [[BFChatRoomEditController alloc]init];
        vc.roomInfo = self.roomInfo;
        [self.navigationController pushViewController:vc animated:YES];
    }
}

- (void)chatRoomTipsDidTap
{
    if (self.roomInfo) {
        BFEditTodInfoController *vc = [[BFEditTodInfoController alloc]init];
        vc.roomInfo = self.roomInfo;
        [self.navigationController pushViewController:vc animated:YES];
    }
}

/// 加入聊天室成功
- (void)enterChatRoom:(NSNotification *)note
{
    self.chatController.roomInfo = self.roomInfo;
    self.navView.navTitleL.text = self.roomInfo.room_name;
}

/// 接收到聊天室事件通知处理
- (void)chatRoomEvent:(NSNotification *)note
{
    MSGroupEvent *event = note.object;
    //事件类型：
      //1：聊天室已被解散
      //2：聊天室属性已修改
      //3：管理员 %s 将本聊天室设为听众模式
      //4: 管理员 %s 恢复聊天室发言功能
      //5：管理员 %s 上线
      //6：管理员 %s 下线
      //7: 管理员 %s 将用户 %s 禁言
      //8: 管理员 %s 将用户 %s、%s 等人禁言
      //9: %s 成为本聊天室管理员
      //10: 管理员 %s 指派 %s 为临时管理员
      //11：管理员 %s 指派 %s、%s 等人为临时管理员
    switch (event.tips.event) {
        case 1:
        {
            [MSHelper showToastString:@"This chatroom is dismissed."];
            [self.navigationController popToRootViewControllerAnimated:YES];
        }
            break;
        case 2:
        {
            [self.chatController.messageController addSystemTips:@"Chatroom info has been changed."];
        }
            break;
        case 3:
        {
            NSString *uid = [NSString stringWithFormat:@"%@",event.tips.uids.firstObject];
            MSProfileInfo *info = [[MSProfileProvider provider]providerProfileFromLocal:uid];
            [self.chatController.messageController addSystemTips:[NSString stringWithFormat:@"%@ disabled sending message for this chatroom.", XMNoNilString(info.nick_name)]];
        }
            break;
        case 4:
        {
            NSString *uid = [NSString stringWithFormat:@"%@",event.tips.uids.firstObject];
            MSProfileInfo *info = [[MSProfileProvider provider]providerProfileFromLocal:uid];
            [self.chatController.messageController addSystemTips:[NSString stringWithFormat:@"%@ enabled sending message for this chatroom.",XMNoNilString(info.nick_name)]];
        }
            break;
        case 5:
        {
            NSString *uid = [NSString stringWithFormat:@"%@",event.tips.uids.firstObject];
            if (![uid isEqualToString:[MSIMTools sharedInstance].user_id]) {
                MSProfileInfo *info = [[MSProfileProvider provider]providerProfileFromLocal:uid];
                [self.chatController.messageController addSystemTips:[NSString stringWithFormat:@"Admin: %@ entered this room.", XMNoNilString(info.nick_name)]];
            }
        }
            break;
        case 6:
        {
            NSString *uid = [NSString stringWithFormat:@"%@",event.tips.uids.firstObject];
            if (![uid isEqualToString:[MSIMTools sharedInstance].user_id]) {
                MSProfileInfo *info = [[MSProfileProvider provider]providerProfileFromLocal:uid];
                [self.chatController.messageController addSystemTips:[NSString stringWithFormat:@" Admin: %@ leaved this room.",XMNoNilString(info.nick_name)]];
            }
        }
            break;
        case 7:
        {
            NSString *uid1 = [NSString stringWithFormat:@"%@",event.tips.uids.firstObject];
            NSString *uid2 = [NSString stringWithFormat:@"%@",event.tips.uids.lastObject];
            MSProfileInfo *info1 = [[MSProfileProvider provider]providerProfileFromLocal:uid1];
            MSProfileInfo *info2 = [[MSProfileProvider provider]providerProfileFromLocal:uid2];
            [self.chatController.messageController addSystemTips:[NSString stringWithFormat:@"%@ muted %@. Reason: %@.",XMNoNilString(info1.nick_name),XMNoNilString(info2.nick_name),XMNoNilString(event.reason)]];
        }
            break;
        case 8:
        {
            if (event.tips.uids.count <= 2) return;
            NSString *managerName;
            NSString *membersStr;
            for (NSInteger i = 0; i < event.tips.uids.count; i++) {
                NSString *uid = [NSString stringWithFormat:@"%@",event.tips.uids[i]];
                if (i == 0) {
                    managerName = [[MSProfileProvider provider]providerProfileFromLocal:uid].nick_name;
                }else {
                    NSString *name = [[MSProfileProvider provider]providerProfileFromLocal:uid].nick_name;
                    if (membersStr == nil) {
                        membersStr = XMNoNilString(name);
                    }else {
                        membersStr = [NSString stringWithFormat:@"%@、%@",membersStr,XMNoNilString(name)];
                    }
                }
            }
            [self.chatController.messageController addSystemTips:[NSString stringWithFormat:@"%@ muted %@. Reason: %@.",XMNoNilString(managerName),XMNoNilString(membersStr),XMNoNilString(event.reason)]];
        }
            break;
        case 9:
        {
            if (event.tips.uids.count < 1) return;
            NSString *name = [[MSProfileProvider provider]providerProfileFromLocal:[NSString stringWithFormat:@"%@",event.tips.uids.firstObject]].nick_name;
            [self.chatController.messageController addSystemTips:[NSString stringWithFormat:@"%@ becomes the admin of this room.",XMNoNilString(name)]];
        }
            break;
        case 10:
        {
            if (event.tips.uids.count < 2) return;
            NSString *adminName = [[MSProfileProvider provider]providerProfileFromLocal:[NSString stringWithFormat:@"%@",event.tips.uids.firstObject]].nick_name;
            NSString *userName = [[MSProfileProvider provider]providerProfileFromLocal:[NSString stringWithFormat:@"%@",event.tips.uids.lastObject]].nick_name;
            [self.chatController.messageController addSystemTips:[NSString stringWithFormat:@"%@ assigned %@ as a temporary admin of the room.",XMNoNilString(adminName),XMNoNilString(userName)]];
        }
            break;
        case 11:
        {
            if (event.tips.uids.count <= 2) return;
            NSString *managerName;
            NSString *membersStr;
            for (NSInteger i = 0; i < event.tips.uids.count; i++) {
                NSString *uid = [NSString stringWithFormat:@"%@",event.tips.uids[i]];
                if (i == 0) {
                    managerName = [[MSProfileProvider provider]providerProfileFromLocal:uid].nick_name;
                }else {
                    NSString *name = [[MSProfileProvider provider]providerProfileFromLocal:uid].nick_name;
                    if (membersStr == nil) {
                        membersStr = XMNoNilString(name);
                    }else {
                        membersStr = [NSString stringWithFormat:@"%@、%@",membersStr,XMNoNilString(name)];
                    }
                }
            }
            [self.chatController.messageController addSystemTips:[NSString stringWithFormat:@"%@ assigned %@ as temporary admins of the room.",XMNoNilString(managerName),XMNoNilString(membersStr)]];
        }
            break;
        default:
            break;
    }
}

#pragma mark - MSChatRoomControllerDelegate

- (void)chatController:(MSChatRoomController *)controller didSendMessage:(MSIMElem *)elem
{
    //主动发送的每一条消息都会进入这个回调，你可以在此做一些统计埋点等工作。。。
}

//将要展示在列表中的每和条消息都会先进入这个回调，你可以在此针对自定义消息构建数据模型
- (MSMessageCellData *)chatController:(MSChatRoomController *)controller prepareForMessage:(MSIMElem *)elem
{
    return nil;
}

- (Class)chatController:(MSChatRoomController *)controller onShowMessageData:(MSMessageCellData *)cellData
{
    //你可以自定义消息气泡的UI,对基本消息类型你可以直接返回nil，采用默认样式
    return nil;
}

///点击某一“更多”单元的回调委托
- (void)chatController:(MSChatRoomController *)controller onSelectMoreCell:(MSInputMoreCell *)cell
{

}

///点击消息头像回调
- (void)chatController:(MSChatRoomController *)controller onSelectMessageAvatar:(MSMessageCell *)cell
{
    NSLog(@"tap avatar...");
}

///点击消息内容回调
- (void)chatController:(MSChatRoomController *)controller onSelectMessageContent:(MSMessageCell *)cell
{
    if (cell.messageData.elem.type == MSIM_MSG_TYPE_IMAGE || cell.messageData.elem.type == MSIM_MSG_TYPE_VIDEO) {
        NSMutableArray *tempArr = [NSMutableArray array];
        NSInteger defaultIndex = 0;
        for (NSInteger i = 0; i < self.chatController.messageController.uiMsgs.count; i++) {
            MSMessageCellData *data =  self.chatController.messageController.uiMsgs[i];
            MSMessageCell *dataCell = (MSMessageCell *)[self.chatController.messageController.tableView cellForRowAtIndexPath:[NSIndexPath indexPathForRow:i inSection:0]];
            if (data.elem.type == MSIM_MSG_TYPE_IMAGE) {
                MSIMImageElem *imageElem = (MSIMImageElem *)data.elem;
                YBIBImageData *imageData = [YBIBImageData new];
                if ([[NSFileManager defaultManager]fileExistsAtPath:imageElem.path]) {
                    imageData.imagePath = imageElem.path;
                }else {
                    imageData.imageURL = [NSURL URLWithString:imageElem.url];
                }
                imageData.projectiveView = dataCell.container.subviews.firstObject;
                [tempArr addObject:imageData];
            }else if (data.elem.type == MSIM_MSG_TYPE_VIDEO) {
                MSIMVideoElem *videoElem = (MSIMVideoElem *)data.elem;
                YBIBVideoData *videoData = [YBIBVideoData new];
                if ([[NSFileManager defaultManager]fileExistsAtPath:videoElem.videoPath]) {
                    videoData.videoURL = [NSURL fileURLWithPath:videoElem.videoPath];
                }else {
                    videoData.videoURL = [NSURL URLWithString:videoElem.videoUrl];
                }
                videoData.projectiveView = dataCell.container.subviews.firstObject;
                [tempArr addObject:videoData];
            }
            if (cell.messageData == data) {
                defaultIndex = tempArr.count-1;
            }
        }
        YBImageBrowser *browser = [YBImageBrowser new];
        browser.dataSourceArray = tempArr;
        browser.currentPage = defaultIndex;
        [browser show];
    }else if (cell.messageData.elem.type == MSIM_MSG_TYPE_LOCATION) {
        
    }
}

@end

File: Swift/MSIM-Swift/ChatDetail/ChatRoom/BFChatRoomViewController.swift
//
//  BFChatRoomViewController.swift
//  MSIM-Swift
//
//  Created by benny wang on 2021/11/2.
//

import UIKit
import MSIMSDK
import ZLPhotoBrowser
import Photos
import Kingfisher


public class BFChatRoomViewController: BFBaseViewController {

    public let chatController: MSChatRoomController = MSChatRoomController()
    
    private var roomInfo: MSGroupInfo? {
        return MSChatRoomManager.sharedInstance().chatroomInfo
    }
    
    public override func viewDidLoad() {
        super.viewDidLoad()

        view.backgroundColor = UIColor.d_color(light: MSMcros.TController_Background_Color, dark: MSMcros.TController_Background_Color_Dark)
        navigationItem.rightBarButtonItem = UIBarButtonItem(barButtonSystemItem: .edit, target: self, action: #selector(editClick))
        chatController.delegate = self
        chatController.roomInfo = self.roomInfo
        addChild(chatController)
        view.addSubview(chatController.view)
        
        self.title = self.roomInfo?.room_name
        addNotifications()
    }

    private func addNotifications() {
        NotificationCenter.default.addObserver(forName: NSNotification.Name.init(MSUIKitNotification_EnterChatroom_success), object: nil, queue: OperationQueue.main) {[weak self] note in
            self?.enterChatRoom()
        }
        NotificationCenter.default.addObserver(forName: NSNotification.Name.init(MSUIKitNotification_ChatRoom_Event), object: nil, queue: OperationQueue.main) {[weak self] note in
            self?.chatRoomEvent(note: note)
        }
    }
    
    /// 加入聊天室成功
    private func enterChatRoom() {
        self.chatController.roomInfo = self.roomInfo
        self.title = self.roomInfo?.room_name
    }
    
    /// 聊天室设置界面
    @objc private func editClick() {
        
        if self.roomInfo != nil {
            self.view.endEditing(true)
            let vc = BFChatRoomEditController()
            vc.roomInfo = self.roomInfo
            navigationController?.pushViewController(vc, animated: true)
        }
    }
    
    /// 接收到聊天室事件通知处理
    @objc func chatRoomEvent(note: Notification) {
        
        //事件类型：
          //1：聊天室已被解散
          //2：聊天室属性已修改
          //3：管理员 %s 将本聊天室设为听众模式
          //4: 管理员 %s 恢复聊天室发言功能
          //5：管理员 %s 上线
          //6：管理员 %s 下线
          //7: 管理员 %s 将用户 %s 禁言
          //8: 管理员 %s 将用户 %s、%s 等人禁言
          //9: %s 成为本聊天室管理员
          //10: 管理员 %s 指派 %s 为临时管理员
          //11：管理员 %s 指派 %s、%s 等人为临时管理员
        guard let event = note.object as? MSGroupEvent,let tips = event.tips,let tipEvent = event.tips?.event else {return}
        
        switch tipEvent {
        case 1:
            MSHelper.showToastWithText(text: "This chatroom is dismissed.")
            navigationController?.popViewController(animated: true)
        case 2:
            chatController.messageController.addSystemTips(text: "Chatroom info has been changed.")
        case 3:
            if let uid = tips.uids.first as? Int,
               let info = MSProfileProvider.shared().providerProfile(fromLocal: String(uid)) {
                chatController.messageController.addSystemTips(text: "\(info.nick_name) disabled sending message for this chatroom.")
            }
        case 4:
            if let uid = tips.uids.first as? Int,
               let info = MSProfileProvider.shared().providerProfile(fromLocal: String(uid)) {
                chatController.messageController.addSystemTips(text: "\(info.nick_name) enabled sending message for this chatroom.")
            }
        case 5:
            if let uid = tips.uids.first as? Int,
               String(uid) != MSIMTools.sharedInstance().user_id,
               let info = MSProfileProvider.shared().providerProfile(fromLocal: String(uid)) {
                chatController.messageController.addSystemTips(text: "Admin: \(info.nick_name) entered this room.")
            }
        case 6:
            if let uid = tips.uids.first as? Int,
               String(uid) != MSIMTools.sharedInstance().user_id,
               let info = MSProfileProvider.shared().providerProfile(fromLocal: String(uid)) {
                chatController.messageController.addSystemTips(text: "Admin: \(info.nick_name) leaved this room.")
            }
        case 7:
            if let uid1 = tips.uids.first as? Int,
               let uid2 = tips.uids.last as? Int,
               let info1 = MSProfileProvider.shared().providerProfile(fromLocal: String(uid1)),
               let info2 = MSProfileProvider.shared().providerProfile(fromLocal: String(uid2)) {
                chatController.messageController.addSystemTips(text: "\(info1.nick_name) muted \(info2.nick_name). Reason: \(event.reason ?? "nothing").")
            }
        case 8:
            if tips.uids.count <= 2 {return}
            var managerName: String = ""
            var membersStr: String = ""
            for (index,uid) in tips.uids.enumerated() {
                if index == 0 {
                    managerName = MSProfileProvider.shared().providerProfile(fromLocal: String(uid as! Int))?.nick_name ?? ""
                }else {
                    let name = MSProfileProvider.shared().providerProfile(fromLocal: String(uid as! Int))?.nick_name ?? ""
                    if membersStr == "" {
                        membersStr = name
                    }else {
                        membersStr = "\(membersStr)、\(name)"
                    }
                }
            }
            chatController.messageController.addSystemTips(text: "\(managerName) muted \(membersStr). Reason: \(event.reason ?? "nothing").")
        case 9:
            if tips.uids.count < 1 {return}
            if let uid = event.tips?.uids.first as? Int,
               let name = MSProfileProvider.shared().providerProfile(fromLocal: String(uid))?.nick_name {
                chatController.messageController.addSystemTips(text: "\(name) becomes the admin of this room.")
            }
        case 10:
            if tips.uids.count < 2 {return}
            if let adminUid = event.tips?.uids.first as? Int,
               let userUid = event.tips?.uids.last as? Int,
               let adminName = MSProfileProvider.shared().providerProfile(fromLocal: String(adminUid))?.nick_name,
               let userName = MSProfileProvider.shared().providerProfile(fromLocal: String(userUid))?.nick_name {
                chatController.messageController.addSystemTips(text: "\(adminName) assigned \(userName) as a temporary admin of the room.")
            }
        case 11:
            if tips.uids.count <= 2 {return}
            var managerName: String = ""
            var membersStr: String = ""
            for (index,uid) in tips.uids.enumerated() {
                if index == 0 {
                    managerName = MSProfileProvider.shared().providerProfile(fromLocal: String(uid as! Int))?.nick_name ?? ""
                }else {
                    let name = MSProfileProvider.shared().providerProfile(fromLocal: String(uid as! Int))?.nick_name ?? ""
                    if membersStr == "" {
                        membersStr = name
                    }else {
                        membersStr = "\(membersStr)、\(name)"
                    }
                }
            }
            chatController.messageController.addSystemTips(text: "\(managerName) assigned \(membersStr) as temporary admins of the room.")
        default:
            break
        }
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self)
    }
    
}

extension BFChatRoomViewController: MSChatRoomControllerDelegate {
    public func didSendMessage(controller: MSChatRoomController, message: MSIMMessage) {
        
    }
    
    public func prepareForMessage(controller: MSChatRoomController, message: MSIMMessage) -> MSMessageCellData? {
        return nil
    }
    
    public func onShowMessageData(controller: MSChatRoomController, cellData: MSMessageCellData) -> MSMessageCell.Type? {
        return nil
    }
    
    public func onSelectMoreCell(controller: MSChatRoomController, cell: MSInputMoreCell) {
        if cell.data?.type == MSIMMoreType.photo {
            selectAsset(isPhoto: true)
            
        }else if cell.data?.type == MSIMMoreType.video {
            selectAsset(isPhoto: false)
        }
    }
    
    public func onSelectMessageAvatar(controller: MSChatRoomController, cell: MSMessageCell) {
        
    }
    
    public func onSelectMessageContent(controller: MSChatRoomController, cell: MSMessageCell) {
        if cell.messageData?.message.type == .MSG_TYPE_IMAGE || cell.messageData?.message.type == .MSG_TYPE_VIDEO {
            //将消息列表中的图片和视频都筛出来
            let tempArr = controller.messageController.uiMsgs.filter { $0.message.type == .MSG_TYPE_IMAGE || $0.message.type == .MSG_TYPE_VIDEO}
            let defaultIndex = tempArr.firstIndex(of: cell.messageData!)
            let lantern = Lantern()
            lantern.pageIndicator = LanternNumberPageIndicator()
            lantern.numberOfItems = {
                tempArr.count
            }
            lantern.cellClassAtIndex = { index in
                let data =  tempArr[index]
                return data.message.type ==  .MSG_TYPE_VIDEO ? VideoZoomCell.self : LanternImageCell.self
            }
            lantern.reloadCellAtIndex = { context in
                let data =  tempArr[context.index]
                if let imageCell = context.cell as? LanternImageCell,let imageElem = data.message.imageElem {
                    
                    if let path = imageElem.path, FileManager.default.fileExists(atPath: path) {
                        imageCell.imageView.kf.setImage(with: URL(fileURLWithPath: path))
                    }else {
                        imageCell.imageView.kf.setImage(with: URL(string: imageElem.url ?? ""))
                    }
                    // 添加长按事件,保存到相册
                    imageCell.longPressedAction = {[weak self] (cell, _) in
                        self?.longPress(elem: imageElem,lantern: cell.lantern)
                    }
                } else if let videoCell = context.cell as? VideoZoomCell,let videoElem = data.message.videoElem {
                    
                    if videoElem.coverImage != nil {
                        videoCell.imageView.image = videoElem.coverImage
                    }else if let coverPath = videoElem.coverPath, FileManager.default.fileExists(atPath: coverPath) {
                        videoCell.imageView.kf.setImage(with: URL(fileURLWithPath: coverPath))
                    }else {
                        videoCell.imageView.kf.setImage(with: URL(string: videoElem.coverUrl ?? "")!)
                    }
                    if let videoPath = videoElem.videoPath, FileManager.default.fileExists(atPath: videoPath) {
                        videoCell.player.replaceCurrentItem(with: AVPlayerItem(url: URL(fileURLWithPath: videoPath)))
                    }else {
                        videoCell.player.replaceCurrentItem(with: AVPlayerItem(url: URL(string: videoElem.videoUrl ?? "")!))
                    }
                    // 添加长按事件,保存到相册
                    videoCell.longPressedAction = {[weak self] (cell, _) in
                        self?.longPress(elem: videoElem,lantern: cell.lantern)
                    }
                }
            }
            
            lantern.cellWillAppear = { cell, index in
                if let videoCell = cell as? VideoZoomCell {
                    videoCell.player.play()
                }
            }
            lantern.cellWillDisappear = { cell, index in
                if let videoCell = cell as? VideoZoomCell {
                    videoCell.player.pause()
                }
            }
            lantern.transitionAnimator = LanternZoomAnimator(previousView: {[weak self] index -> UIView? in
                
                let data = tempArr[index]
                if let atIndex = self?.chatController.messageController.uiMsgs.firstIndex(of: data) {
                    let indexPath = IndexPath(row: atIndex, section: 0)
                    let cell = controller.messageController.tableView.cellForRow(at: indexPath) as? MSMessageCell
                    return cell?.container.subviews.first
                }
                return nil
            })
            lantern.pageIndex = defaultIndex ?? 0
            lantern.show()
        }
    }
}

extension BFChatRoomViewController {
    
    private func selectAsset(isPhoto: Bool) {
        let config = ZLPhotoConfiguration.default()
        config.allowSelectVideo = !isPhoto
        config.allowSelectImage = isPhoto
        config.maxSelectCount = 1
        config.allowEditImage = isPhoto
        config.allowEditVideo = !isPhoto
        let editConfig = ZLEditImageConfiguration()
        editConfig.imageStickerContainerView = ImageStickerContainerView()
        config.editImageConfiguration = editConfig
        
        config.canSelectAsset = { (asset) -> Bool in
            return true
        }
        
        config.noAuthorityCallback = { (type) in
            switch type {
            case .library:
                debugPrint("No library authority")
            case .camera:
                debugPrint("No camera authority")
            case .microphone:
                debugPrint("No microphone authority")
            }
        }
        
        let picker = ZLPhotoPreviewSheet(selectedAssets: nil)
        picker.selectImageBlock = {[weak self] (images,assets,isOriginal) in
            self?.didPickerAsset(images: images,assets: assets,isPhoto: isPhoto)
        }
        picker.cancelBlock = {
            
        }
        picker.selectImageRequestErrorBlock = { (errorAssets, errorIndexs) in
            debugPrint("fetch error assets: \(errorAssets), error indexs: \(errorIndexs)")
        }
        picker.showPhotoLibrary(sender: self)
    }
    
    private func didPickerAsset(images: [UIImage],assets: [PHAsset],isPhoto: Bool) {
        if isPhoto {
            for (index,image) in images.enumerated() {
                
                let imagePath = FileManager.pathForIMImage() + NSString.uuid() + ".jpg"
                let imageData = image.pngData()
                try? imageData?.write(to: URL(fileURLWithPath: imagePath))
                let message = MSIMManager.sharedInstance().createImageMessage(imagePath, identifierID: assets[index].localIdentifier)
                self.chatController.sendMessage(message: message)
            }
        }else {
            MSHelper.showToast()
            var count: Int = 0
            for (index,video) in assets.enumerated() {
                
                _ = ZLPhotoManager.fetchAVAsset(forVideo: video) { avasset, info in
                    
                    let localPath = FileManager.pathForIMVideo() + "\(NSString.uuid()).mp4"
                    let exportSession = AVAssetExportSession(asset: avasset!, presetName: AVAssetExportPresetPassthrough)
                    exportSession?.outputURL = URL(fileURLWithPath: localPath)
                    exportSession?.outputFileType = .mp4
                    exportSession?.exportAsynchronously {
                        switch exportSession?.status {
                        case .completed:
                            count += 1
                            let coverPath = FileManager.pathForIMVideo() + NSString.uuid() + ".jpg"
                            let coverData = images[index].pngData()
                            try? coverData?.write(to: URL(fileURLWithPath: coverPath))
                            let message = MSIMManager.sharedInstance().createVideoMessage(localPath, type: "mp4", duration: Int(video.duration), snapshotPath: coverPath, identifierID: nil)
                            self.chatController.sendMessage(message: message)
                            
                        case .failed, .cancelled:
                            count += 1
                        default:
                            print(exportSession?.error ?? "")
                        }
                        if count == assets.count {
                            MSHelper.dismissToast()
                        }
                    }
                }
            }
        }
    }
    
    /// 长按图片或视频，保存到相册
    private func longPress(elem: MSIMElem,lantern: Lantern?) {
        let alert = UIAlertController(title: nil, message: nil, preferredStyle: .actionSheet)
        alert.addAction(UIAlertAction(title: "Save to album", style: .default, handler: {[weak self] _ in
            if let imageElem = elem as? MSIMImageElem {
                self?.savePhotoToAlbum(elem: imageElem)
            }else if let videoElem = elem as? MSIMVideoElem {
                self?.saveVideoToAlbum(elem: videoElem)
            }
            
        }))
        alert.addAction(UIAlertAction(title: "Cancel", style: .cancel, handler: nil))
        lantern?.present(alert, animated: true, completion: nil)
    }
    
    private func savePhotoToAlbum(elem: MSIMImageElem) {
        var saveImage: UIImage?
        if let image = elem.image {
            saveImage = image
        }else if let path = elem.path,FileManager.default.fileExists(atPath: path) {
            saveImage = UIImage.init(contentsOfFile: path)
        }else if let url = elem.url,let data = try? Data.init(contentsOf: URL(string: url)!) {
            saveImage = UIImage.init(data: data)
        }
        if saveImage != nil {
            ZLPhotoManager.saveImageToAlbum(image: saveImage!) { isOK, _ in
                if isOK {
                    MSHelper.showToastSuccWithText(text: "Done")
                }else {
                    MSHelper.showToastFailWithText(text: "Error")
                }
            }
        }
    }
    
    private func saveVideoToAlbum(elem: MSIMVideoElem) {
        
        //1.如果视频本地文件存在，直接保存到相册
        if let path = elem.videoPath,FileManager.default.fileExists(atPath: path) {
            ZLPhotoManager.saveVideoToAlbum(url: URL(fileURLWithPath: path)) { isOK, _ in
                if isOK {
                    MSHelper.showToastSuccWithText(text: "Done")
                }else {
                    MSHelper.showToastFailWithText(text: "Error")
                }
            }
        }else if let url = elem.videoUrl {
            //2.如果视频本地文件不存在，需要先下载，再保存到相册
            MSHelper.showToast()
            let tempUrl = "\(FileManager.pathForIMVideo())" + NSString.uuid() + ".mp4"
            MSIMManager.sharedInstance().uploadMediator?.ms_download?(fromUrl: url, toSavePath: tempUrl, progress: { progress in
                MSHelper.showProgress(progress: Float(progress), text: "Video is downloading...")
            }, succ: { _ in
                
                ZLPhotoManager.saveVideoToAlbum(url: URL(fileURLWithPath: tempUrl)) { isOK, _ in
                    if isOK {
                        MSHelper.showToastSuccWithText(text: "Done")
                    }else {
                        MSHelper.showToastFailWithText(text: "Error")
                    }
                    try? FileManager.default.removeItem(atPath: tempUrl)
                }
                
            }, fail: { code, errorString in
                
                MSHelper.showToastFailWithText(text: errorString)
            })
        }
    }
}
